pennylane_lightning/src/simulator/KernelMap.hpp

```diff
@@ -0,0 +1,322 @@
+    static inline const int dummy = assignDefaultKernelsForMatrixOp();
+};
+} // namespace Internal
+///@endcond
+
+///@cond DEV
+struct DispatchElement {
+    uint32_t priority;
+    Util::IntegerInterval<size_t> interval;
+    Gates::KernelType kernel;
+};
+
+inline bool lower_priority(const DispatchElement &lhs,
+                           const DispatchElement &rhs) {
+    return lhs.priority < rhs.priority;
+}
+
+inline bool higher_priority(const DispatchElement &lhs,
+                            const DispatchElement &rhs) {
+    return lhs.priority > rhs.priority;
+}
+
+/**
+ * @brief Maintain dispatch element using a vector decreasingly-ordered by
+ * priority.
+ */
+class PriorityDispatchSet {
+  private:
+    std::vector<DispatchElement> ordered_vec_;
+
+  public:
+    [[nodiscard]] bool
+    conflict(uint32_t test_priority,
+             const Util::IntegerInterval<size_t> &test_interval) const {
+        const auto test_elt = DispatchElement{test_priority, test_interval,
+                                              Gates::KernelType::None};
+        const auto [b, e] =
+            std::equal_range(ordered_vec_.begin(), ordered_vec_.end(), test_elt,
+                             higher_priority);
+        for (auto iter = b; iter != e; ++iter) {
+            if (!is_disjoint(iter->interval, test_interval)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    void insert(const DispatchElement &elt) {
+        const auto iter_to_insert = std::upper_bound(
+            ordered_vec_.begin(), ordered_vec_.end(), elt, &higher_priority);
+        ordered_vec_.insert(iter_to_insert, elt);
+    }
+
+    template <typename... Ts> void emplace(Ts &&...args) {
+        auto elt = DispatchElement{std::forward<Ts>(args)...};
+        const auto iter_to_insert = std::upper_bound(
+            ordered_vec_.begin(), ordered_vec_.end(), elt, &higher_priority);
+        ordered_vec_.insert(iter_to_insert, elt);
+    }
+
+    [[nodiscard]] Gates::KernelType getKernel(size_t num_qubits) const {
+        for (const auto &elt : ordered_vec_) {
+            if (elt.interval(num_qubits)) {
+                return elt.kernel;
+            }
+        }
+        PL_ABORT("Cannot find a kernel for the given number of qubits.");
+    }
+
+    void clearPriority(uint32_t remove_priority) {
+        const auto begin = std::lower_bound(
+            ordered_vec_.begin(), ordered_vec_.end(), remove_priority,
+            [](const auto &elt, uint32_t p) { return elt.priority > p; });
+        const auto end = std::upper_bound(
+            ordered_vec_.begin(), ordered_vec_.end(), remove_priority,
+            [](uint32_t p, const auto &elt) { return p > elt.priority; });
+        ordered_vec_.erase(begin, end);
+    }
+};
+///@endcond
+
+struct AllThreading {};
+struct AllMemoryModel {};
+
+constexpr static AllThreading all_threading{};
+constexpr static AllMemoryModel all_memory_model{};
```

Comment:
May be worth adding some info for these structs if we are declaring them within the Doxygen generation scope